### PR TITLE
feat: composable workflows and orchestrator execution model

### LINKED ISSUE #20: Hierarchical Workflow Management with Rollback Support
### Feature to Enhance

Enhance the workflow engine to support hierarchical workflows, allowing addition and deletion of sub-workflows (each containing steps), sequential execution, and robust rollback on error.

## Sub-Workflow Management
- Ability to add sub-workflows within a parent workflow.
- Each sub-workflow can contain multiple steps.

## Sequential Execution
- Executing a workflow should run all its sub-workflows in defined order.
- Each sub-workflow executes its steps sequentially.

## Rollback on Error
If any step or sub-workflow fails, initiate rollback.
- Rollback all sub-workflows in reverse order.
- Each sub-workflow rolls back its steps in reverse order.

## Acceptance Criteria:
- API or interface to build workflows with sub-workflows.
- Execution engine runs sub-workflows and their steps in order.
- On error, rollback is triggered as described.
- Rollback operations are idempotent and handle partial failures gracefully.

### Business Need

This feature will provide flexible, reliable workflow orchestration with strong error recovery, supporting complex automation scenarios.

### Additional Context

_No response_

### LINKED ISSUE #22: Replace Choreography Execution Model with Orchestrator for Centralized Workflow Control
Currently, the workflow engine uses a choreography execution model, where each step is responsible for invoking the next step or triggering rollback. This decentralized approach can make it difficult to enforce business rules, handle errors consistently, and monitor workflow progress.

### Proposal:
Replace the choreography execution model with an orchestrator execution model. In this model, a central orchestrator manages the execution order, error handling, and rollback of all steps and sub-workflows.

### Benefits:

- Centralized Control: Easier to enforce business logic and manage execution flow.
- Consistent Error Handling: Centralized error, retry, and compensation logic.
- Simplified Composition: Easier to compose and nest workflows.
- Improved Observability: Single point for logging, reporting, and monitoring.

### Acceptance Criteria:

- The orchestrator (aka `Workflow` execution) manages all step and sub-workflows execution.
- Error handling and rollback are performed centrally.
- Existing tests pass; new tests cover orchestrator scenarios.
- Documentation and examples are updated to reflect the new model.

### Additional Context:
This change will enable more robust, maintainable, and observable workflow execution, supporting future extensibility and advanced features.

### Plan of Action
- Add `Step` interface to `Workflows` in order to support #20 
- Implement `Forward` and `Reverse` methods of `Step` interface for `Workflows` to implement orchestrator execution model

### PR DESCRIPTION
**Pull request by leninmehedy**
## Description
This PR introduces support for composable workflows and replaces choreography with orchestrator execution model to the automa package.

### Composable Workflows:
- Workflows can now be composed of other workflows and steps, enabling modular and reusable workflow definitions.
- The Workflow interface extends Step, allowing workflows to be nested within other workflows.
- Added tests to verify execution order and correctness of composed workflows.

### Orchestrator Execution Model:
- Introduced an orchestrator pattern for workflow execution, enabling centralized control over step execution and rollback. This simplifies execution of steps and workflows with simpler logic and debugging capability. See orchestrator vs choreography execution model differences here: https://stackoverflow.com/questions/4127241/orchestration-vs-choreography
- Enhanced error handling and reporting for orchestrated workflows.

### API Improvements:
- Updated Registry and Workflow interfaces to support step/workflow composition.
- Improved documentation and comments for public interfaces.

### Improved test coverage
Screenshot 2025-08-07 at 8 09 25 pm

### Related Issues

* Closes #20 
* Closes: #22

* Related to #

### Pull request (PR) checklist

* \[x] This PR added tests (unit, integration, and/or end-to-end)
* \[ ] This PR updated documentation
* \[ ] This PR added no TODOs or commented out code
* \[ ] This PR has no breaking changes
* \[ ] Any technical debt has been documented as a separate issue and linked to this PR
* \[ ] Any `go.mod` and/or `go.sum` changes have been explained to and approved by a repository manager
* \[ ] All related issues have been linked to this PR
* \[ ] All changes in this PR are included in the description
* \[ ] When this PR merges the commits will be squashed and the title will be used as the commit message, the 'commit
  message guidelines' below have been followed

### Testing

* \[ ] This PR added unit tests
* \[ ] This PR added integration/end-to-end tests
* \[ ] These changes required manual testing that is documented below
* \[ ] Anything not tested is documented

The following manual testing was done:

* TBD

The following was not tested:

* TBD

<details>
<summary>
Commit message guidelines
</summary>
We use 'Conventional Commits' to ensure that our commit messages are easy to read, follow a consistent format, and for automated release note generation. Please follow the guidelines below when writing your commit messages:

1. BREAKING CHANGE: a commit that has a footer BREAKING CHANGE:, or appends a ! after the type/scope, introduces a
   breaking API change (correlating with MAJOR in Semantic Versioning). A BREAKING CHANGE can be part of commits of any
   type. NOTE: currently breaking changes will only bump the MAJOR version.
2. The title is prefixed with one of the following:

| Prefix    | Description                                         | Semantic Version Update | Captured in Release Notes |
|-----------|-----------------------------------------------------|-------------------------|---------------------------|
| feat:     | a new feature                                       | MINOR                   | Yes                       |
| fix:      | a bug fix                                           | PATCH                   | Yes                       |
| perf:     | performance                                         | PATCH                   | Yes                       |
| refactor: | code change that isn't feature or fix               | none                    | No                        |
| test:     | adding missing tests                                | none                    | No                        |
| docs:     | changes to documentation                            | none                    | Yes                       |
| build:    | changes to build process                            | none                    | No                        |
| ci:       | changes to CI configuration                         | none                    | No                        |
| style:    | formatting, missing semi-colons, etc                | none                    | No                        |
| chore:    | updating grunt tasks etc; no production code change | none                    | No                        |

</details>

### CODE REVIEW

File: workflow.go [hunk 1]
```diff
@@ -11,131 +12,240 @@ var nolog = zerolog.Nop()
...
 
 // GetStepSequence returns the ordered list of Step IDs in the workflow.
 // Returns a copy to prevent external modification.
 func (wf *workflow) GetStepSequence() []string {
-	var stepSequence []string
-	for step := wf.first; step != nil; step = step.GetNext() {
-		stepSequence = append(stepSequence, step.GetID())
-	}
-	return stepSequence
+	return wf.stepSequence
```
[the remark this answers is not in the document]
Comment (leninmehedy) [reply workflow.go:225]: updated old doc

**Comment by leninmehedy** [comment]
Superseded by: https://github.com/automa-saga/automa/pull/32
